### ISSUE TITLE
WIP: Fixing the error in test/constraints.jl

### ISSUE THREAD

**Issue by IainNZ** [JuliaOpt/JuMP.jl#1429]
The problematic line is
`cref = @constraint(m, [x, 2x] .== [1-x, 3])`

I've verified parsing hasn't change between 0.6 and 0.7, and I think 1.0 is fine too.
I did that with macroexpand, which also can can be cleaned up to get to this generated code.

```
using JuMP
import MathOptInterface
m = Model()
@variable(m, x)
(JuMP.validmodel)(m, :m)
foo = (JuMP.Val){false}()
bar = (JuMP.destructive_add_with_reorder!)(foo, [x])
fizz = (JuMP.destructive_add_with_reorder!)(bar, -1.0, [1 - x])
output = (JuMP.addconstraint).(
            m,
            (JuMP.buildconstraint).(
                ()->error(),
                fizz,
                MathOptInterface.EqualTo{Float64}(0.0)))
println(output)
```

The inner broadcast is problematic at the least - the outer one may be too.

From https://github.com/JuliaLang/julia/blob/release-0.7/NEWS.md, the relevant text:

> Previously, broadcast defaulted to treating its arguments as scalars if they were not arrays. This behavior is deprecated, and in the future broadcast will default to iterating over all its arguments. Wrap arguments you wish to be treated as scalars with Ref() or a 1-tuple. Package developers can choose to allow a non-iterable type T to always behave as a scalar by implementing broadcastable(x::T) = Ref(x) (#26212).

**Comment by IainNZ** [comment]
The suggested fix does work:
```
output = (JuMP.addconstraint).(
            Ref(m),
            (JuMP.buildconstraint).(
                Ref(()->error()),
                fizz,
                Ref(MathOptInterface.EqualTo{Float64}(0.0))))
```
But I need to dig deeper into the code to understand what the best solution is

**Comment by blegat** [comment]
Shouldn't we just implement broadcastable for MOI sets and JuMP models ?

**Comment by mlubin** [comment]
Ref https://github.com/JuliaOpt/MathOptInterface.jl/pull/472

**Comment by IainNZ** [comment]
Feels kind of gross, but I suppose that works.

**Comment by mlubin** [comment]
Implementing broadcastable makes sense for JuMP models. It seems a bit sketchy for MOI sets since these are objects that have dimension of some sort.

**Comment by blegat** [comment]
> It seems a bit sketchy for MOI sets since these are objects that have dimension of some sort.

It is consistent with the `eachscalar` MOIU function that turns an AbstractVectorFunction into something interable. The MOI sets and functions are not iterable but `eachscalar` turns them into iterable objects

**Comment by IainNZ** [comment]
Closed by #1438
